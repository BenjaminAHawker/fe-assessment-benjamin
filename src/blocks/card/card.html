<!DOCTYPE html>
<html lang="en">
    <head>
        <title>Card Block</title>
    </head>
    <body>
        <h1>Card Block</h1>
        <div id="app"></div>
    </body>
</html>