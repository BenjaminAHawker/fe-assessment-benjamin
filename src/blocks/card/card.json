{
  "title": "Powerlines",
  "body": "Here's some awesome descriptive text about powerlines! The point of this text is to test responsive styling.",
  "image": {
    "src": "https://www.wesco.com/content/dam/wesco/assets/wesco/images/600/Power_Lines_Orange_Sunset_1316185179_RGB.jpg",
    "alt": "powerlines"
  },
  "button": {
    "href": "#",
    "label": "Learn more"
  }
}