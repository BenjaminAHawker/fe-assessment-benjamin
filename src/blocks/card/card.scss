$text-color: #808080;

h1,h2,h3{
  padding:0;
  margin:0;
};

.card {
  max-width: 32rem;
  margin-bottom: 20px;
  padding: 20px;
  display: flex;
  flex-direction: column;
  justify-content: center;

  .imageFull {
    border-radius: 5px;
  }
  .header {
    color: $text-color;
    font-size: 32px;
    margin: .5em 0;
  }
  .information {
    color: $text-color;
    font-size: 18px;
    font-weight: 500;
  }
  .cta {
    width: 100%;
    display: flex;
    justify-content: center;
    text-decoration: none;
    color: white;
    background-color: #033042;
    border-radius: 5px;
    padding: .5em;
    font-weight: 800;
  }
}