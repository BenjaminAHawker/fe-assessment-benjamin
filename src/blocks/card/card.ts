import { ICard } from '../../types/Card';
import data from '../card/card.json';
import '../card/card.scss';

/**
 * Card decorator function
 *
 * This function accepts props defined in types/Card.ts and
 * marshals in data from the card.json file located in this
 * file's directory.
 *
 * From here, we would look to target the "app" div defined
 * in our "card.html" file where we will now be able to append
 * a parent (wrapper) element and using all the child elements
 * this component comprises to create a re-usable component.
 *
 * @param {ICard} props - Card Object
 *
 */

const decorate = (props:ICard):void => {
    const app: HTMLElement | null = document.getElementById("app");
    const parent = document.createElement("div");
    parent.classList.add('card');

    // Page Header Text
    const h1 = document.createElement("h1");
    h1.textContent = "Card Block";
    parent.append(h1);

    const { title, body, image, button } = props;

    // Image Block

    const imgBlock = document.createElement("img");
    imgBlock.classList.add('imageFull');
    imgBlock.src = image?.src || '';
    imgBlock.alt = image?.alt || '';
    parent.appendChild(imgBlock);

    // Text Block
    const textContainer = document.createElement('div');

    const header = document.createElement('h1');
    header.classList.add('header')
    header.innerHTML = `${title || ''}`;

    const informationText = document.createElement('p');
    informationText.classList.add('information')
    informationText.innerHTML = `${body || ''}`

    // Button Block
    const btn = document.createElement('a');
    btn.classList.add('cta');
    btn.innerHTML = `<span>${button?.label}</span>`;
    btn.href = button?.href || '';

    // Render Items
    textContainer.appendChild(header);
    textContainer.appendChild(informationText);
    textContainer.appendChild(btn);
    parent.appendChild(textContainer);


    try{
        app?.append(parent);
    }catch(error){
        console.error(error);
    }
}

decorate(data);
