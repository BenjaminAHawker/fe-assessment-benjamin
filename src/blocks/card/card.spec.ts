import * as fs from "fs";
import * as path from "path";
import data from './card.json';
const html = fs.readFileSync(path.resolve(__dirname, './card.html'));

jest.dontMock('fs');

describe('Card ->', function () {
    beforeEach(function() {
        document.documentElement.innerHTML = html.toString();
        require('./card');
    });

    afterEach(function() {
        jest.resetModules();
    });

    it('should have access to DOM', () => {
        expect(document.getElementById('app')).toBeTruthy();
    });

    describe('Element Render Checks', () => {

        it('should render image',() => {
            expect(document.getElementsByClassName('imageFull')).toBeTruthy();
        });

        it('should render correct text',() => {
            const header = document.getElementsByClassName('header');
            expect(header).toMatchSnapshot();
        });

        it('should render button' ,() => {
            expect(document.getElementsByClassName('cta')).toBeTruthy();
        })

    })
});