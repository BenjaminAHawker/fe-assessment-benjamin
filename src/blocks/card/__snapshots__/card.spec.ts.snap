// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card -> Element Render Checks should render correct text 1`] = `
HTMLCollection [
  <h1
    class="header"
  >
    Powerlines
  </h1>,
]
`;
